pxd_fastpath.c

```diff
@@ -869,6 +869,35 @@ static void pxd_process_io(struct pxd_io_tracker *head)
 	}
 }
 
+// external request to suspend IO on fastpath device
+int pxd_request_suspend(struct pxd_device *pxd_dev, bool skip_flush)
+{
+	struct pxd_fastpath_extension *fp = &pxd_dev->fp;
+	int nfd = fp->nfd;
+	int i;
+
+	if (pxd_dev->using_blkque || !nfd || !fp->fastpath || fp->app_suspend) {
+		return -EINVAL;
+	}
+
+	fp->app_suspend = true;
+	pxd_suspend_io(pxd_dev);
+
+	if (skip_flush) return 0;
+
+	for (i = 0; i < nfd; i++) {
+		if (fp->file[i] > 0) {
+			int ret = vfs_fsync(fp->file[i], 0);
```

Comment:
This needs to be an ioctl. A single unresponsive device can cause px-storage to be stuck in D state.

Reply:
I did some experiments to understand this better. The conclusion is, it does not matter what system call interface one would use to enter kernel mode, be it a read/write or a ioctl, the D state is caused by actions  from within the call. If the processing path, puts in uninterruptible sleep, then process state switches to D state, i.e. it does not matter whether read/write or ioctl interface, if process waits in  uninterruptible mode, then process state is D (disk sleep). The only option is to take the processing vfs_sync() out of the call and do it in the background and make userspace check this again for completion. I will take this up as a secondary task out of this PR.

Here is the kernel module:
```
root@bionic:/home/lns/srcs/src/kmodule# cat testmod.c
#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/blkdev.h>
#include <linux/miscdevice.h>
#include <linux/jiffies.h>
#include <linux/sched.h>
#include <linux/uio.h>

struct testheader {
        int cmd;
};

static
long testmod_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
{
        switch (cmd) {
        case 0:
                printk("testmod_ioctl interruptible sleep\n");
                // sleep for 60 secs
                schedule_timeout_interruptible(msecs_to_jiffies(60000));
                break;
        case 1:
                printk("testmod_ioctl uninterruptible sleep\n");
                // sleep for 60 secs
                schedule_timeout_uninterruptible(msecs_to_jiffies(60000));
                break;
        default:
                return -EINVAL;
        }
        return 0;
}

static
ssize_t testmod_read_iter(struct kiocb *iocb, struct iov_iter *iter)
{
        //struct file *file = iocb->ki_filp;
        struct testheader th;
        size_t len = sizeof(th);

        printk("testmod_read_iter iter.count %ld", iter->count);
        if (copy_from_iter(&th, len, iter) != len) {
                printk("testmod_read_iter cannot copy header");
                return -EFAULT;
        }

        printk("testmod_read_iter header cmd %d\n", th.cmd);
        switch (th.cmd) {
        case 0:
                // sleep for 60 secs
                schedule_timeout_interruptible(msecs_to_jiffies(60000));
                break;
        case 1:
                // sleep for 60 secs
                schedule_timeout_uninterruptible(msecs_to_jiffies(60000));
                break;
        default:
                return -EINVAL;
        }

        return iter->count;
}

ssize_t testmod_write_iter(struct kiocb *iocb, struct iov_iter *iter)
{
        //struct file *file = iocb->ki_filp;
        struct testheader th;
        size_t len = sizeof(th);

        if (copy_from_iter(&th, len, iter) != len) {
                return -EFAULT;
        }

        switch (th.cmd) {
        case 0:
                // sleep for 60 secs
                schedule_timeout_interruptible(msecs_to_jiffies(60000));
                break;
        case 1:
                // sleep for 60 secs
                schedule_timeout_uninterruptible(msecs_to_jiffies(60000));
                break;
        default:
                return -EINVAL;
        }

        return iter->count;
}

static const struct file_operations fops = {
        .owner = THIS_MODULE,
        .read_iter = testmod_read_iter,
        .write_iter = testmod_write_iter,
        .unlocked_ioctl = testmod_ioctl,
};

struct testcontext {
        struct miscdevice misc;
        bool ok;
};

static int testctxt_init(struct testcontext *tctx)
{
        int err;
        tctx->misc.minor = MISC_DYNAMIC_MINOR;
        tctx->misc.name = "testcontrol";
        tctx->misc.fops = &fops;

        err =  misc_register(&tctx->misc);
        if (err) {
                printk(KERN_ERR "testctxt: failed to register misc device %s: err %d\n",
                                tctx->misc.name, err);
        }
        tctx->ok = (err == 0);
        return err;
}

static void testctxt_exit(struct testcontext *tctx)
{
        if (tctx->ok) {
                misc_deregister(&tctx->misc);
        }
}

static struct testcontext tctxt;

int testmod_entry(void)
{
        return testctxt_init(&tctxt);
}

void testmod_exit(void)
{
        testctxt_exit(&tctxt);
}

module_init(testmod_entry);
module_exit(testmod_exit);

MODULE_LICENSE("GPL");
MODULE_VERSION("0.1");
```
Here is the userspace code:
```
root@bionic:/home/lns/srcs/src/umodule# cat testmain.c
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <assert.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/uio.h>

#define TC "/dev/testcontrol"

#define CMD_INTR (0)
#define CMD_UNINTR (1)

int test_ioctl(int fd, long cmd)
{
        int rc = ioctl(fd, cmd);

        printf("iocl interface, cmd %ld, rc = %d\n", cmd, rc);
        return rc;
}

struct testheader {
        int cmd;
};

int test_read(int fd, long cmd)
{
        struct testheader th = { cmd };
        struct iovec iov = { &th, sizeof(th)};
        int rc = readv(fd, &iov, 1);

        printf("read interface, cmd %ld, rc = %d\n", cmd, rc);
        return rc;
}

int main(int argc, char *argv[])
{
        int fd = open(TC, O_RDWR);
        char ch;

        assert(fd > 0);

        // use ioctl
        /*
        printf("trying with ioctl interface..... using interruptible cmd\n");
        test_ioctl(fd, CMD_INTR);
        printf("completed ioctl interface..... using interruptible cmd\n");
        getchar();
        */

        printf("trying with ioctl interface..... using uninterruptible cmd\n");
        test_ioctl(fd, CMD_UNINTR);
        printf("completed ioctl interface..... using uninterruptible cmd\n");
        getchar();

        // use read/write interface
        /*
        printf("trying with read interface..... using interruptible cmd\n");
        test_read(fd, CMD_INTR);
        printf("completed read interface..... using interruptible cmd\n");
        getchar();
        */

        printf("trying with read interface..... using uninterruptible cmd\n");
        test_read(fd, CMD_UNINTR);
        printf("completed read interface..... using uninterruptible cmd\n");

        close(fd);

        return 0;
}
```

Reply:
will take this up as a subsequent revision.